daffodil-test/src/test/resources/org/apache/daffodil/section23/dfdl_expressions/expressions3.tdml

```diff
@@ -0,0 +1,400 @@
+    <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd" />
+    <dfdl:format ref="ex:GeneralFormat" lengthKind="delimited"/>
+
+    <xs:group name="g">
+      <!-- 
+        The path expression below is depending on the ability to 
+        compile polymorphically. The expression ../foo/hdr is always an int.
+        The expression ../foo/bar is a string, or a complex type depending on 
+        the context. So this shouldn't compile since you can't do math on 
+        a complex type. 
+       -->
+      <xs:sequence>
+      <xs:choice dfdl:choiceDispatchKey="{ 
+         xs:string(if (../foo/hdr eq 1) 
+           then xs:int( ../foo/bar )
+           else xs:int( ../foo/bar + 1 ) )}">
+        <xs:element name="c1" type="xs:string" dfdl:choiceBranchKey="1" />
+        <xs:element name="c2" type="xs:string" dfdl:choiceBranchKey="2" />
+      </xs:choice>
+      </xs:sequence>
+    </xs:group>
+
+    <xs:element name="r">
+      <xs:complexType>
+        <xs:sequence dfdl:separator="|">
+          <xs:element name="e1">
+            <xs:complexType>
+              <xs:sequence>
+                <xs:element name="foo">
+                  <xs:complexType>
+                    <xs:sequence dfdl:separator="|">
+                      <xs:element name="hdr" type="xs:int"/>
+                      <xs:element name="bar" type="xs:float" />
+                      <xs:group ref="ex:g" />
+                    </xs:sequence>
+                  </xs:complexType>
+                </xs:element>
+              </xs:sequence>
+            </xs:complexType>
+          </xs:element>
+          <xs:element name="e2">
+            <xs:complexType>
+              <xs:sequence>
+                <xs:element name="foo">
+                  <xs:complexType>
+                    <xs:sequence dfdl:separator="|">
+                      <xs:element name="hdr" type="xs:int"/>
+                      <xs:element name="bar">
+                        <xs:complexType>
+                          <xs:sequence>
+                            <xs:element name="quux" type="xs:int" />
+                          </xs:sequence>
+                        </xs:complexType>
+                      </xs:element>
+                      <xs:group ref="ex:g" />
+                    </xs:sequence>
+                  </xs:complexType>
+                </xs:element>
+              </xs:sequence>
+            </xs:complexType>
+          </xs:element>
+        </xs:sequence>
+      </xs:complexType>
+    </xs:element>
+ </tdml:defineSchema>
+
+  <tdml:parserTestCase name="test_polymorphic_expr_3" model="s3" root="r"
+    description="Polymorphic path expression">
+    <tdml:document>1|1|a|2|1|b</tdml:document>
+    <tdml:errors>
+      <tdml:error>Schema Definition Error</tdml:error>
+      <tdml:error>cannot be converted to xs:int</tdml:error>
+      <tdml:error>Left operand for operator '+' must have numeric type</tdml:error>
```

Comment:
What does the context look like for this error? Ideally, we would have both: the location where the expression is defined (eg. the lexical context of g), and the location where its usage causes a problem (eg. e1/foo)

Reply:
We should add this to the test; as this seems like a place where it would be easy to accidentally break our diagnostics without noticing.

Reply:
Added checks that two different contexts for the different types, are exhibited.